File: worker/_shared/moment-inserts.js
/** Per-line moment illustrations (full-frame inserts) — edge port of server/images/moment_inserts.py */

import { freemiumExtract } from "./freemium-extract.js";
import { normalizeExpressionBucket } from "./expression-bucket.js";

const EXPRESSION_PROMPTS = {
  sad: "sad wistful expression, downturned eyes, soft melancholy",
  angry: "angry fierce expression, narrowed eyes, tense jaw",
  whisper: "quiet secretive expression, softened lips, intent gaze",
  yell: "shouting intense expression, open mouth, emphatic",
  happy: "bright happy smile, lively eyes",
  surprised: "surprised wide eyes, startled expression",
  scared: "fearful wide-eyed expression, tense shoulders, recoiling",
  excited: "exhilarated grin, bright energetic eyes",
  embarrassed: "flustered blushing expression, averted gaze",
  smug: "smug half-smile, confident raised eyebrow",
  tender: "soft affectionate expression, warm gentle eyes",
  nervous: "anxious fidgety expression, uncertain eyes",
  sarcastic: "dry deadpan expression, one eyebrow raised",
  determined: "resolute focused expression, set jaw",
  desperate: "desperate pleading expression, strained eyes",
};

export function lineExpression(line) {
  return normalizeExpressionBucket(line?.expression);
}

export function expressionPromptSuffix(expression) {
  return EXPRESSION_PROMPTS[expression] || `${expression} facial expression`;
}

/** @returns {{ scene, lineIndex, line } | null} */
export function lineAtIndex(analysis, lineIdx) {
  let idx = 0;
  for (const scene of analysis.scenes || []) {
    for (let li = 0; li < (scene.lines || []).length; li += 1) {
      if (idx === lineIdx) {
        return { scene, lineIndex: li, line: scene.lines[li] };
      }
      idx += 1;
    }
  }
  return null;
}

export function momentDescription(analysis, scene, line, { lineIdx = -1 } = {}) {
  const custom = line?.moment_prompt || "";
  if (String(custom).trim()) return String(custom).trim();
  const byId = Object.fromEntries((analysis.characters || []).map((c) => [c.id, c]));
  const cid = line?.character_id;
  const char = byId[cid];
  const name = char?.name || (cid === "narrator" ? "Narrator" : cid);
  const expr = lineExpression(line);
  const exprBit = expr !== "normal" ? expressionPromptSuffix(expr) : "dramatic expressive moment";
  const loc = scene?.location || scene?.title || "scene";
  const text = String(line?.text || "").trim().slice(0, 200);

  const cast = (scene?.present_character_ids || [cid].filter(Boolean))
    .map((id) => {
      const c = byId[id];
      const label = c?.name || id;
      const look = c?.description ? `: ${String(c.description).slice(0, 120)}` : "";
      return `${label}${look}`;
    })
    .join("; ");

  const appearance = char?.description
    ? `Keep ${name} visually consistent — ${String(char.description).slice(0, 160)}. `
    : "";

  return (
    `${loc}. Full-screen story moment. Characters present: ${cast || name}. `
    + `${appearance}${name}, ${exprBit}. Scene: ${scene?.title || scene?.id}. `
    + `Story beat: ${text}`
  ).trim();
}

export function patchAnalysisLine(analysis, lineIdx, line) {
  const loc = lineAtIndex(analysis, lineIdx);
  if (!loc) return analysis;
  const { scene, lineIndex } = loc;
  const scenes = (analysis.scenes || []).map((s) => {
    if (s.id !== scene.id) return s;
    const newLines = [...(s.lines || [])];
    newLines[lineIndex] = line;
    return { ...s, lines: newLines };
  });
  return { ...analysis, scenes };
}

export async function tweakMomentLine(analysis, scene, line, { useLlm = false, env } = {}) {
  const updates = { visual_moment: true };
  if (!line?.moment_prompt) {
    updates.moment_prompt = momentDescription(analysis, scene, line);
  }
  let next = { ...line, ...updates };

  const disabled = String(env?.DISABLE_MOMENT_SCRIPT_TWEAK || "").toLowerCase();
  if (!useLlm || ["1", "true", "yes"].includes(disabled)) {
    return next;
  }

  try {
    const byId = Object.fromEntries((analysis.characters || []).map((c) => [c.id, c]));
    const char = byId[line.character_id];
    const system = (
      "You polish visual-audiobook moment tags. Return JSON only: "
      + '{"moment_prompt": "image gen description", "text": "optional refined line"}'
    );
    const user = (
      `Scene: ${scene.title}. Location: ${scene.location}.\n`
      + `Character: ${char?.name || line.character_id}.\n`
      + `Line (${line.kind}): ${line.text}\n`
      + `Expression: ${line.expression}\n`
      + "Write a vivid moment_prompt for a full-screen illustration (fan-service OK when "
      + "the line warrants it). Keep text identical unless a tiny clarity fix helps."
    );
    const result = await freemiumExtract(user, { systemPrompt: system, env });
    let data = result?.data || {};
    if (typeof data === "string") {
      data = JSON.parse(data.replace(/,(\s*[}\]])/g, "$1"));
    }
    if (data.moment_prompt) updates.moment_prompt = String(data.moment_prompt).trim();
    if (data.text && String(data.text).trim()) {
      const newText = String(data.text).trim();
      if (newText !== String(line.text || "").trim()) updates.text = newText;
    }
    next = { ...line, ...updates };
  } catch {
    /* LLM polish optional */
  }
  return next;
}

export function stableInsertSeed(key) {
  let h = 0;
  const s = `insert:${key}`;
  for (let i = 0; i < s.length; i += 1) h = (h * 31 + s.charCodeAt(i)) >>> 0;
  return h % 2147483647;
}

File: legacy/server/images/moment_inserts.py
"""Per-line moment illustrations (full-frame inserts)."""
from __future__ import annotations

import json
import os
import re
from pathlib import Path

from ..analyze.schema import AnalysisLine, AnalysisScene, BookAnalysis
from .expression_sprites import _line_expression, expression_prompt_suffix


def line_at_index(analysis: BookAnalysis, line_idx: int) -> tuple[AnalysisScene, int, AnalysisLine] | None:
    idx = 0
    for scene in analysis.scenes:
        for li, line in enumerate(scene.lines):
            if idx == line_idx:
                return scene, li, line
            idx += 1
    return None


def moment_description(
    analysis: BookAnalysis,
    scene: AnalysisScene,
    line: AnalysisLine,
    *,
    line_idx: int,
) -> str:
    custom = getattr(line, "moment_prompt", None) or ""
    if custom.strip():
        return custom.strip()
    by_id = {c.id: c for c in analysis.characters}
    cid = line.character_id
    char = by_id.get(cid)
    name = char.name if char else (cid if cid != "narrator" else "Narrator")
    expr = _line_expression(line)
    expr_bit = expression_prompt_suffix(expr) if expr != "normal" else "dramatic expressive moment"
    loc = scene.location or scene.title or "scene"
    text = (line.text or "").strip()[:200]
    return (
        f"{loc}. Full-screen story moment: {name}, {expr_bit}. "
        f"Scene: {scene.title or scene.id}. Story beat: {text}"
    ).strip()


def tweak_moment_line(
    analysis: BookAnalysis,
    scene: AnalysisScene,
    line: AnalysisLine,
    *,
    use_llm: bool = False,
) -> AnalysisLine:
    """Mark line as a visual moment; optionally enrich prompt / polish tag via LLM."""
    updates: dict = {"visual_moment": True}
    if not getattr(line, "moment_prompt", None):
        updates["moment_prompt"] = moment_description(
            analysis, scene, line, line_idx=-1,
        )
    if not use_llm or os.environ.get("DISABLE_MOMENT_SCRIPT_TWEAK", "").lower() in (
        "1", "true", "yes",
    ):
        return line.model_copy(update=updates)

    try:
        from ..analyze.freemium_extract import freemium_extract
        by_id = {c.id: c for c in analysis.characters}
        char = by_id.get(line.character_id)
        system = (
            "You polish visual-audiobook moment tags. Return JSON only: "
            '{"moment_prompt": "image gen description", "text": "optional refined line"}'
        )
        user = (
            f"Scene: {scene.title}. Location: {scene.location}.\n"
            f"Character: {char.name if char else line.character_id}.\n"
            f"Line ({line.kind}): {line.text}\n"
            f"Expression: {line.expression}\n"
            "Write a vivid moment_prompt for a full-screen illustration (fan-service OK when "
            "the line warrants it). Keep text identical unless a tiny clarity fix helps."
        )
        result = freemium_extract(user, system_prompt=system)
        data = result.get("data") or {}
        if isinstance(data, str):
            data = json.loads(re.sub(r",(\s*[}\]])", r"\1", data))
        if data.get("moment_prompt"):
            updates["moment_prompt"] = str(data["moment_prompt"]).strip()
        if data.get("text") and str(data["text"]).strip():
            new_text = str(data["text"]).strip()
            if new_text != (line.text or "").strip():
                updates["text"] = new_text
    except Exception:
        pass
    return line.model_copy(update=updates)


def patch_analysis_line(analysis: BookAnalysis, line_idx: int, line: AnalysisLine) -> BookAnalysis:
    loc = line_at_index(analysis, line_idx)
    if not loc:
        return analysis
    scene, li, _ = loc
    new_lines = list(scene.lines)
    new_lines[li] = line
    scenes = []
    for s in analysis.scenes:
        if s.id == scene.id:
            scenes.append(s.model_copy(update={"lines": new_lines}))
        else:
            scenes.append(s)
    return analysis.model_copy(update={"scenes": scenes})


def collect_visual_inserts(analysis: BookAnalysis) -> list[dict]:
    """All moment lines that need generated inserts (no cap)."""
    from .expression_sprites import INSERT_EXPRESSIONS  # noqa: PLC0415

    by_id = {c.id: c for c in analysis.characters}
    primary = {c.id for c in analysis.characters if c.importance == "primary"}
    inserts: list[dict] = []
    line_idx = 0
    for scene in analysis.scenes:
        for line in scene.lines:
            cid = line.character_id
            expr = _line_expression(line)
            visual = bool(getattr(line, "visual_moment", False))
            has_epub = getattr(line, "illustration_ref", None) is not None
            if not visual and not has_epub:
                if (
                    cid in primary
                    and line.kind == "dialogue"
                    and expr in INSERT_EXPRESSIONS
                    and float(getattr(line, "intensity", 1.0) or 1.0) >= 0.75
                    and len((line.text or "").split()) >= 4
                ):
                    visual = True
            if visual and not has_epub:
                inserts.append({
                    "line_idx": line_idx,
                    "character_id": cid,
                    "scene_id": scene.id,
                    "expression": expr,
                    "description": moment_description(
                        analysis, scene, line, line_idx=line_idx,
                    ),
                })
            line_idx += 1
    return inserts


def reference_bytes_for_moment(
    analysis: BookAnalysis,
    character_id: str,
    style_dir: Path,
    reference_images: list[bytes] | None,
) -> list[bytes] | None:
    from ..playback.illustrations import reference_bytes_for_character

    p = style_dir / f"char_{character_id}.png"
    if p.is_file():
        try:
            return [p.read_bytes()]
        except OSError:
            pass
    return reference_bytes_for_character(character_id, analysis, reference_images)
